markdown: Add support for inline audio files playback

roanster007 opened the pull request:
**This PR adds the backend markdown for supporting playback of inline audio files without leaving the browser.**

Fixes: [#27007](https://github.com/zulip/zulip/issues/27007) Enable playback of audio files.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

![JNhjyZCH0g](https://github.com/zulip/zulip/assets/97049524/ce003b31-37b1-4217-9d7a-8e7151f5e5cb)


- Currently added support only for ```mp3``` (mpeg) and ```aac``` audio formats only.
- Tested Audio from [link](https://file-examples.com/index.php/sample-audio-files/sample-mp3-download/).

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

roanster007 commented:
@alya could you have a look at this PR once you're free, please.

alya commented:
Nice! A few thoughts/questions:

1. What file formats does this support? (I tested with an mp3.)
2. The volume control is confusing, as it appears on top of the playback line, and only contrasts a bit with it. Can you suggest a different approach? Please feel free to post in #design on CZO for feedback.
<img width="435" alt="Screenshot 2023-10-31 at 9 12 06 AM" src="https://github.com/zulip/zulip/assets/2090066/f686303b-3cb2-41a7-ad54-a186ba9ffc4a">

4. This download tooltip incorrectly appears when hovering over the volume control, where clicking won't cause the file to be downloaded:
<img width="433" alt="Screenshot 2023-10-31 at 9 12 21 AM" src="https://github.com/zulip/zulip/assets/2090066/c677ab34-9b72-4b5c-b639-ed60b5e0f24d">

5. The mouse pointer is a hand everywhere, including over the time area. But clicking on the time has no effect. Should it be a regular arrow pointer there?

roanster007 commented:
@alya so,

1. I have currently added support for .mp3 and .aac file formats. (I could've added more, but needed some feedback regarding what all formats to consider).
2. Regarding the volume control and all, I have added the default controls that come with the ```controls``` attribute of the  ```<audio>``` tag. So, it's not that flexible regarding all those UI/UX and all.

Should I add a custom controls for the Audio?
Since this issue was just about adding this support for the audio files playback, I didn't do anything out of it.

roanster007 commented:
@alya ,

- Regarding the file formats supported, I have added for .mp3, and .aac for now, and am ready to add more based on the suggestions.
- Regarding the Volume Control, since we are using default audio controls, there is pretty much nothing we can do regarding that since it is dependent on how Browsers render it, and I couldn't find a way to override it.
- Regarding the remaining 2 points, I have updated the code, to serve the needs.

Regarding the design for a custom audio control, as per the [CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/UI.20for.20audio.20controls/near/1672451), a new issue can be raised to update the UI.

timabbott commented on zerver/lib/markdown/__init__.py:
```diff
@@ -572,6 +572,36 @@ def run(self, root: Element) -> None:
+
+    This rewrites all audio, except ones that are served from the current
+    realm or global STATIC_URL. This is to ensure that each realm only loads
+    audios that are hosted on that realm or by the global installation,
```
s/audios/audio files/.

timabbott commented on zerver/lib/markdown/__init__.py:
```diff
@@ -572,6 +572,36 @@ def run(self, root: Element) -> None:
+            if is_static_or_current_realm_url(url, self.zmd.zulip_realm):
+                continue
+            audio.set("src", get_camo_url(url))
+            audio.set("data-audio-original-url", url)
```
To what extent does this function differ from the similar one for images?

roanster007 replied:
@timabbott , this function is rather similar to the ```InlineVideoProcessor``` one, and I have picked it up, since both show nearly the same behaviour while looking up for elements in the Tree and setting the urls.
The images one just has ```camo url``` since there the support is for upload from device, while here we're dealing with upload from links as well, so the original audio url is also included.

roanster007 replied:
OK looking at it now, it would be better if I refactored `InlineVideoProcessor` to also manage the audio files.

Updated it to `InlineMediaProcessor`, which manages both video and audio files

alya commented:
I started a discussion about what audio formats to allow: https://chat.zulip.org/#narrow/stream/101-design/topic/audio.20formats/near/1675024

roanster007 commented:
@timabbott , I have update the comment suggested, and have updated the audio files supported as suggested in the [CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/audio.20formats/near/1675024).

roanster007 commented:
@timabbott , I have updated the comment suggested, and have updated the audio files supported as suggested in the [CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/audio.20formats/near/1675024). 
CI failed at Test-Locked-Requirements due to the outage.

timabbott commented on zerver/tests/test_link_embed.py:
```diff
@@ -667,31 +667,38 @@ def test_link_preview_non_html_data(self) -> None:
-        )
+        msg.refresh_from_db()
+        expected_content = f"""<div class="message_inline_audio"><a href="{url}"><audio controls data-audio-original-url="{url}" preload="metadata" src="{get_camo_url(url)}" title="{url}"></audio></a></div>"""
+        self.assertEqual(expected_content, msg.rendered_content)
```
So I think we need to not remove the preexisting test, but instead modify it and add a new test. 

I believe we still support via the `FetchLinksEmbedData` code path, handling audio objects (I guess you can test one with a different file extension but an audio content-type, so that the new code doesn't apply)

roanster007 replied:
@timabbott , I have tried leaving it, but it was throwing me an error of mocks not called ([link](https://chat.zulip.org/#narrow/stream/49-development-help/topic/backend.20test.20failed.20in.20CI.20for.20inline.20playback.20of.20audio.20files/near/1671205)). I checked if the problem is my code or the test by adjusting the test to replicate it on video (.mp4) elements rather (which also has embed data), and the same test fails.
Hence, I had to modify the test, as per the test used for the YouTube Links rendering

roanster007 replied:
Updated!

timabbott commented:
This won't make the 8.0 release, so I've labeled it as something to review just after the release.

terpimost commented:
I suggest to make audio player of the full width of the container, so it would act as a block element and not as inline

alya commented:
I don't fully understand the terminology... Is that consistent with allowing message that combine audio with other content?

roanster007 commented on zerver/tests/test_link_embed.py:
```diff
@@ -652,38 +652,6 @@ def test_invalid_link(self) -> None:
-    def test_link_preview_non_html_data(self) -> None:
-        user = self.example_user("hamlet")
-        self.login_user(user)
-        url = "http://test.org/audio.mp3"
```
Removed this test since `audio` is no longe a non html element. Added test for `<audio>` tags rather in `markdown_test_cases.json`

timabbott replied:
I think based on the test name, it might be intending to test any file type that we don't support? So you might just want to update it to use a file type that is not `mp3`.

roanster007 replied:
makes sense, I updated it to a `.wav` format, which is not supported

timabbott commented on web/src/markdown.ts:
```diff
@@ -26,7 +26,7 @@ const preview_regexes = [
     // Inline image and video previews, check for contiguous chars ending in image and video suffix
```
This command can probably be updated to "Inline media previews"?

roanster007 replied:
updated!

timabbott commented on zerver/lib/markdown/__init__.py:
```diff
@@ -552,11 +552,11 @@ def run(self, root: Element) -> None:
 
 
-class InlineVideoProcessor(markdown.treeprocessors.Treeprocessor):
+class InlineMediaProcessor(markdown.treeprocessors.Treeprocessor):
```
`MediaTreeProcessor` might be a better name; I'm not sure the "inline" part is helpful/accurate. 

(We can't change the `inline_image` type names, since they're in the API, but these are typically block elements, not inline elements, in the CSS terminology sense)

timabbott replied:
And renaming `InlineImageProcessor` to `ImageTreeProcessor` might also be appropriate.

roanster007 replied:
Updated `InlineMediaProcessor` to `MediaSourceTreeProcessor`

timabbott commented on zerver/lib/markdown/__init__.py:
```diff
@@ -572,18 +572,18 @@ def __init__(self, zmd: "ZulipMarkdown") -> None:
+            url = media.get("src")
             assert url is not None
             if is_static_or_current_realm_url(url, self.zmd.zulip_realm):
                 # Don't rewrite videos on our own site (e.g. user uploads).
```
s/videos/media/` here.

roanster007 replied:
updated!

timabbott commented on zerver/lib/markdown/__init__.py:
```diff
@@ -1259,23 +1266,34 @@ def add_video(
+        if self.is_video(url):
+            class_attr = "message_inline_image message_inline_video"
+        else:
+            class_attr = "message_inline_audio"
```
We should have an API design conversation on this detail of what the HTML looks like.

timabbott replied:
https://chat.zulip.org/#narrow/channel/368-sponsor-integration/topic/open.20collective/with/2142968

roanster007 replied:
I assume you meant [#api design > audio previews](https://chat.zulip.org/#narrow/channel/378-api-design/topic/audio.20previews/with/2143921)?

I updated it to the form:
```html
<audio class="message_inline_audio" controls preload="metadata" src="https://external-content.zulipcdn.net/external_content/..."></audio>
```
as per the discussion there

timabbott commented on zerver/tests/fixtures/markdown_test_cases.json:
```diff
@@ -471,6 +471,19 @@
     },
+    {
+      "name": "only_inline_audio",
+      "input": "http://test.org/audio.mp3",
```
Use example.com, which is the RFC standard invalid domain, not random other domains like `test.org`, that might be real.

roanster007 replied:
updated!

roanster007 commented:
documented audio and video preview html syntax in `api_docs/message-formatting.md`

timabbott commented on api_docs/message-formatting.md:
````diff
@@ -249,6 +249,16 @@ generate a video preview element with the following format.
+``` html
+<audio class="media_container_audio" controls preload="metadata" src="/user_uploads/path/to/audio.mp3">
+</audio>
+```
````
I tweaked this to be better documentation -- not doing an overly-long line, and documenting the `camo` case.

``` diff
diff --git a/api_docs/message-formatting.md b/api_docs/message-formatting.md
index 1f16822874..f7aeb6db75 100644
--- a/api_docs/message-formatting.md
+++ b/api_docs/message-formatting.md
@@ -257,14 +257,27 @@ generate a video preview element with the following format.
 
 ## Audio previews
 
-When a Zulip message is sent linking to an uploaded audio, Zulip will
-generate an audio preview element with the following format.
+When a Zulip message is sent linking to an uploaded audio file, Zulip
+will generate an HTML audio player element with the following format.
 
 ``` html
-<audio class="media_container_audio" controls preload="metadata" src="/user_uploads/path/to/audio.mp3">
+<audio class="media_container_audio" controls preload="metadata"
+  src="/user_uploads/path/to/audio.mp3">
 </audio>
 ```
 
+If the Zulip server has rewritten the URL of the audio file, it will
+provide the URL in a `data-original-url` parameter.
+
+``` html
+<audio class="media_container_audio" controls preload="metadata"
+  data-original-url="https://example.com/path/to/original/file.mp3"
+  src="https://example.com/path/to/playable/file.mp3">
+</audio>
+```
+
+**Changes**: New in Zulip 11.0 (feature level ZF-59bb5b).
+
 ## Mentions and silent mentions
 
 Zulip markup supports [mentioning](/help/mention-a-user-or-group)
diff --git a/api_docs/unmerged.d/ZF-59bb5b.md b/api_docs/unmerged.d/ZF-59bb5b.md
index e69de29bb2..fd36f61509 100644
--- a/api_docs/unmerged.d/ZF-59bb5b.md
+++ b/api_docs/unmerged.d/ZF-59bb5b.md
@@ -0,0 +1,2 @@
+* [Message formatting](/api/message-formatting): Added new HTML
+  formatting for uploaded audio files generating a player experience.
```

timabbott replied:
And I think we want something like this for the `data-original-url` logic:

``` diff
diff --git a/zerver/lib/markdown/__init__.py b/zerver/lib/markdown/__init__.py
index 74633d6162..3a16ba9083 100644
--- a/zerver/lib/markdown/__init__.py
+++ b/zerver/lib/markdown/__init__.py
@@ -579,11 +579,18 @@ class MediaSourceTreeProcessor(markdown.treeprocessors.Treeprocessor):
             if is_static_or_current_realm_url(url, self.zmd.zulip_realm):
                 # Don't rewrite media on our own site (e.g. user uploads).
                 continue
-            # The href= is still set to the non-Camo'd version, which
-            # is used by clients to play the full media without
-            # imposing load on the Camo server; this src is used to
-            # load the initial thumbnail and metadata (through Camo)
-            media.set("src", get_camo_url(url))
+
+            if e.tag == "video":
+                # The href= is still set to the non-Camo'd version, which
+                # is used by clients to play the full media without
+                # imposing load on the Camo server; this src is used to
+                # load the initial thumbnail and metadata (through Camo)
+                media.set("src", get_camo_url(url))
+            elif e.tag == "audio":
+                media.set("data-original-url", url)
+                media.set("src", get_camo_url(url))
+            else:
+                raise AssertionError("Unexpected media tag!")
 
 
 class BacktickInlineProcessor(markdown.inlinepatterns.BacktickInlineProcessor):
```

timabbott commented on api_docs/message-formatting.md:
````diff
@@ -241,6 +241,20 @@ assume that messages sent prior to the introduction of thumbnailing
+    </video>
+  </a>
+</div>
+```
````
I merged the prep commit as acb12fa953d5ec2b69730c04bf4e1072f1383b11.

timabbott reviewed (commented):
Merged the prep commit.

roanster007 commented:
closing in favor of #34903
